moshi/moshi/models/tts.py

```diff
@@ -0,0 +1,624 @@
+                state.consumption_times.append(step)
+                if entry.tokens:
+                    consumed_new_word = True
+                    state.transcript.append((entry.word, step))
+                    # We queue the tokens to be fed to the model.
+                    state.queued.extend(entry.tokens)
+                    if self.second_stream_ahead:
+                        # We queue the tokens for the N+lookahead word into the second text stream.
+                        state.lookahead.extend(state.get_tokens_ahead(self.second_stream_ahead))
+                    # Entry contains a new word, we reset the max padding counter.
+                    state.remaining_padding = self.max_padding
+                else:
+                    # Entry is only here to insert a break, pretend the token was a PAD.
+                    token = self.token_ids.pad
+                state.forced_padding = entry.padding
+            else:
+                token = self.token_ids.pad
+                if self.second_stream_ahead and state.end_step is None:
+                    token = self.token_ids.word
+                # Trying to consume past the last word, we reached the end.
+                if state.end_step is None:
+                    state.end_step = step
+
+        output: int | None = None
+        if token == self.token_ids.pad:
+            # Decrement the counters for remaining and forced pads.
+            if state.remaining_padding > 0:
+                state.remaining_padding -= 1
+            if state.forced_padding > 0:
+                state.forced_padding -= 1
+            if state.queued:
+                # We have some text tokens to feed to the model.
+                output = state.queued.popleft()
+            else:
+                output = self.token_ids.pad
+        elif token == self.token_ids.word:
+            output = self.token_ids.word
+        elif token == self.token_ids.zero:
+            output = token
+        else:
+            raise RuntimeError(f"Invalid token {token}")
+
+        if self.second_stream_ahead:
+            second = -1
+            if output == self.token_ids.word:
+                # If sampled the `word` special token, we put it on the
+                # second text stream instead of the main one.
+                second = self.token_ids.word
+                if state.queued:
+                    # This allows us to pass the current word tokens faster.
+                    output = state.queued.popleft()
+                else:
+                    output = self.token_ids.pad
+            elif state.lookahead:
+                # Otherwise if we have some lookahead tokens we feed them.
+                second = state.lookahead.popleft()
+            # Then we multiplex the two tokens. We add `+1` to `second` so that
+            # we can encode -1, which would translate to an all 0s embedding.
+            output = (second + 1) * self.token_ids.card + output
```

Comment:
More of a modeling question here but why do we want to have the "dense" cross-product of the two tokens, wouldn't that necessitate a very large embedding table?